File: 0688. Knight Probability in Chessboard/Solution0688.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    double find(int n, int k, int r, int c, vector<vector<vector<double>>>& dp) {
        
        if (r < 0 || r >= n || c < 0 || c >= n) return 0.0;
        if (k == 0) return 1.0;
        if (dp[r][c][k] != 0) return dp[r][c][k];
        double p = find(n, k-1, r-2, c+1, dp) + find(n, k-1, r-1, c+2, dp) + 
                   find(n, k-1, r+1, c+2, dp) + find(n, k-1, r+2, c+1, dp) + 
                   find(n, k-1, r+2, c-1, dp) + find(n, k-1, r+1, c-2, dp) + 
                   find(n, k-1, r-1, c-2, dp) + find(n, k-1, r-2, c-1, dp);
        p /= 8.0;
        dp[r][c][k] = p;
        return p;
    }
    
    double knightProbability(int n, int k, int row, int column) {
        vector<vector<vector<double>>> dp(n, vector<vector<double>>(n, vector<double>(k+1))); // dp[n][n][k+1]
        return find(n, k, row, column, dp);
    }
};

File: 0688. Knight Probability in Chessboard/Solution0688.java
class Solution0688 {
    int[][] dirs = new int[][]{{-1,-2},{-1,2},{1,-2},{1,2},{-2,1},{-2,-1},{2,1},{2,-1}};
    public double knightProbability(int n, int k, int row, int column) {
        double[][][] f = new double[n][n][k + 1];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                f[i][j][0] = 1;
            }
        }
        for (int p = 1; p <= k; p++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    for (int[] d : dirs) {
                        int nx = i - d[0], ny = j - d[1];
                        if (nx < 0 || nx >= n || ny < 0 || ny >= n) continue;
                        f[i][j][p] += f[nx][ny][p - 1] / 8;
                    }
                }
            }
        }
        return f[row][column][k];
    }
}

// 作者：AC_OIer
// 链接：https://leetcode.cn/problems/knight-probability-in-chessboard/solution/gong-shui-san-xie-jian-dan-qu-jian-dp-yu-st8l/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。